### PR TITLE
Fix alignment checking in loading memory arg immediates

### PR DESCRIPTION
**Pull request by guojidan**
closed: #3482

### CODE REVIEW

**Comment by juntao** [comment]
Hello, I am a [code review agent](https://github.com/flows-network/github-pr-review/) on [flows.network](https://flows.network/).

It could take a few minutes for me to analyze this PR. Relax, grab a cup of coffee and check back later. Thanks!

**Review by dannypsnl — DISMISSED** [review]

**Comment by hydai** [comment]
@q82419 PTAL. Does this meet WASM specifications?

**Comment by hydai** [comment]
@guojidan 
The linter check failed. Please fix it.

**Comment by q82419** [comment]
Sorry that in my opinion this fix maybe not so reasonable.
But I should verify it in detail.

**Comment by guojidan** [comment]
Yes, this fix is not correct. Change the PR state to draft.

**Comment by guojidan** [comment]
@q82419 please take a look

**Comment by q82419** [comment]
> @q82419 please take a look

Please pass the tests. Thanks!

File: include/common/enum.inc [hunk 1]
```diff
@@ -881,6 +881,9 @@ E(SharedMemoryNoMax, 0x011A, "shared memory must have maximum")
 E(IntrinsicsTableNotFound, 0x011B, "intrinsics table not found")
 // Malformed table (Ref-types proposal)
 E(MalformedTable, 0x011C, "malformed table")
+// The store alignment must be less than 128 and multi-memory
```
Comment (q82419) [inline include/common/enum.inc:884]: I think the comment can be compressed into 1 line:

`// Alignment must < 64 without and < 128 with multi-memory proposal.`

File: lib/loader/ast/instruction.cpp [hunk 2]
```diff
@@ -313,8 +313,13 @@ Expect<void> Loader::loadInstruction(AST::Instruction &Instr) {
     if (auto Res = readU32(Instr.getMemoryAlign()); unlikely(!Res)) {
       return Unexpect(Res);
     }
-    if (Conf.hasProposal(Proposal::MultiMemories) &&
-        Instr.getMemoryAlign() >= 64) {
+    if (Instr.getMemoryAlign() >= 128 ||
```
Comment (q82419) [inline lib/loader/ast/instruction.cpp:316]: The following is better:

```
if (Align >= 128) {
  ... error
} else if (Align >= 64) {
  if (hasMultiMemories) {
    ... read memory index
  } else {
    ... error
  }
}
```

**Review by q82419 — CHANGES_REQUESTED** [review]
For passing the test, you can modify the `leb128` encoded value in `test/loader/instructionTest.cpp:703` into: `0x8FU, 0x80U, 0x80U, 0x80U, 0x00U`.

File: lib/loader/ast/instruction.cpp (hunk 2 above)
Comment (guojidan) [reply lib/loader/ast/instruction.cpp:316]: but clang suggested do not use 'else' after 'return' 🤔

Comment (dannypsnl) [reply lib/loader/ast/instruction.cpp:316]: That's ok, we don't follow that rule.

**Comment by guojidan** [comment]
Maybe it is necessary to re-run the failed check.

**Comment by q82419** [comment]
Please fix with the requested changes. Thanks!

**Comment by guojidan** [comment]
> Please fix with the requested changes. Thanks!

Fixed 😄

**Review by q82419 — APPROVED** [review]

**Pull request merged**